### PR TITLE
Use OIDC Token directly when options in Service Connections is enabled

### PR DESCRIPTION
**Pull request by ROunofF**
## Description

Add and option to use an OIDC token provided by Azure Devops to get temporary credentials via Assume Role with Web Identity API.

## Motivation

This prevent having to maintain static credentials when running outside of AWS.

## Related Issue(s), If Filed

#521 

## Testing

Pipelines using OIDC for federation into AWS : DownloadS3, the AWSShellScript task, the AWSPowerShellModuleScript, the AWS CLI task.

## Checklist

-   [X] I have read the **README** document
-   [X] I have read the **CONTRIBUTING** document
-   [X] My code follows the code style of this project
-   [] I have added tests to cover my changes
-   [X] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lib/awsConnectionParameters.ts [hunk 1]
```diff
@@ -394,3 +467,30 @@ export async function getRegion(): Promise<string> {
...
+        throw new Error('System.AccessToken is undefined')
+    }
+
+    const authHandler = getHandlerFromToken(token)
+    const connection = new WebApi(uri, authHandler)
+    const api: ITaskApi = await connection.getTaskApi()
+    const response = await api.createOidcToken({}, projectId, hub, planId, jobId, connectedService)
+    if (response == undefined || response.oidcToken == undefined) {
+        throw new Error('Impossible to generate OidcToken')
+    }
```
Comment (rbbarad) [inline src/lib/awsConnectionParameters.ts:489]:
```suggestion
    if (response === undefined || response.oidcToken === undefined) {
        throw new Error('Failed to generate OidcToken')
    }
```

File: package.json [hunk 2]
```diff
@@ -80,9 +80,11 @@
         "validator": "^13.7.0"
     },
     "dependencies": {
+        "@aws-sdk/client-sts": "^3.616.0",
```
Comment (rbbarad) [inline package.json:83]: What is this being used for?

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "aws-vsts-tools",
-    "version": "1.14.0",
+    "version": "1.17.29",
```
[the remark this answers is not in the document]
Comment (ROunofF) [reply package.json]: Fixed

File: src/lib/awsConnectionParameters.ts [hunk 4]
```diff
@@ -394,3 +467,30 @@ export async function getRegion(): Promise<string> {
...
+    const connection = new WebApi(uri, authHandler)
+    const api: ITaskApi = await connection.getTaskApi()
+    const response = await api.createOidcToken({}, projectId, hub, planId, jobId, connectedService)
+    if (response === undefined || response.oidcToken === undefined) {
+        throw new Error('Failed to generate OidcToken')
+    }
+
+    const claims = JSON.parse(Buffer.from(response.oidcToken.split('.')[1], 'base64').toString('utf-8'))
+
+    console.debug('OIDC Token generated: issuer: {%s} sub: {%s}, aud: {%s}', claims.iss, claims.sub, claims.aud)
```
Comment (rbbarad) [inline src/lib/awsConnectionParameters.ts:493]:
```suggestion
    console.log('OIDC Token generated: issuer: {%s} sub: {%s}, aud: {%s}', claims.iss, claims.sub, claims.aud)
```

File: src/lib/awsConnectionParameters.ts [hunk 5]
```diff
@@ -148,6 +152,11 @@ async function assumeRoleFromInstanceProfile(
         return undefined
     }
     const authInfo = getEndpointAuthInfo(awsparams, endpointName)
+
+    if (authInfo.useOIDC === 'true') {
+        console.debug('Skipping Instance profile, we have OIDC enabled')
```
Comment (rbbarad) [inline src/lib/awsConnectionParameters.ts:157]:
```suggestion
        console.log('Skipping Instance profile, we have OIDC enabled')
```

File: src/lib/awsConnectionParameters.ts [hunk 6]
```diff
@@ -166,6 +175,60 @@ async function assumeRoleFromInstanceProfile(
...
+    endpointName: string | undefined
+): Promise<AWS.Credentials | undefined> {
+    if (!endpointName) {
+        return undefined
+    }
+
+    try {
+        const authInfo = getEndpointAuthInfo(awsParams, endpointName)
+        if (!authInfo.useOIDC) {
+            console.debug('Skipping OIDC: not enabled in service connections')
```
Comment (rbbarad) [inline src/lib/awsConnectionParameters.ts:189]:
```suggestion
            console.log('Skipping OIDC: not enabled in service connections')
```

**Review by rbbarad — APPROVED** [review]

**Pull request merged**